packages/SwingSet/test/test-liveslots.js

```diff
@@ -1495,3 +1495,103 @@ test('unserializable promise rejection', async t => {
 
   t.deepEqual(l2.resolutions[0], [expectedPA, true, expectedError]);
 });
+
+test('result promise in args', async t => {
+  const { log, syscall } = buildSyscall();
+  const vatlog = [];
+
+  // A message whose arguments references its own result promise might
+  // cause problems with translation ordering. Liveslots cannot
+  // directly create these (the arguments are hardened by the
+  // E().foo() proxy handler) before userspace is given the result
+  // promise. But it must be able to handle an inbound message in this
+  // shape.
+
+  // For liveslots, this is basically the same as what we get if
+  // msg1.args references a promise which later appears as
+  // msg2.result, which could happen because msg2 was queued in a
+  // kernel promise and fell behind msg1. And it is similar to
+  // msg3.result appearing in a later msg4.args .
+
+  // In both cases, liveslots currently handles this ok, but the
+  // sequence is a bit weird. We'll document it here:
+
+  // * when the promise is imported (in msg.args), we create a Promise
+  //   object, track it in slotToVal/valToSlot, and hold the resolve /
+  //   reject functions in importedPromisesByPromiseID . Userspace
+  //   sees this Promise. A `dispatch.notify` will resolve it. Any
+  //   messages sent to it will be queued into the kernel, which will
+  //   wait for a `syscall.resolve` and then send the messages to the
+  //   resulting object's home vat.
+
+  // * Each dispatch.deliver causes the creation of a `res` Promise
+  //   for the result of the local delivery. If a result= vpid was
+  //   provided, we use `res.then` to wait for it to resolve. If/when
+  //   that happens, `thenHandler` does both a syscall.resolve() and
+  //   checks/resolves `importedPromisesByPromiseID`.
+
+  // * So if we wind up with both a real "imported" Promise and a
+  //   result vpid, any messages our userspace sends to the Promise
+  //   will be queued into the kernel, even though we're going to be
+  //   calling syscall.resolve . If we resolve it to something local,
+  //   those messages will take a round-trip through the kernel. It
+  //   might be nice to forward `res` to the generated Promise, to
+  //   avoid that trip in some cases.
```

Comment:
To clarify my understanding, if we have the mirror flow, where the vat receives as argument the result promise of a previously delivered send, does liveslot realize this and locally queues messages sent to the promise, only forwarding them to the kernel if we resolve to a remote object? I assume the day we add promise redirects, we'll need to forward on redirection as well?

Reply:
It would be nice if liveslots could recognize that case and send fewer messages into the kernel.

What happens now is that the first message (`dispatch.deliver(method='one', result=vpid1)`) creates a Promise and registers it with slotToVal/valToSlot. The Promise actually comes from `HandledPromise.applyMethod()`, as the `returnedP` argument to the handler. The second mesage (`dispatch.deliver(method='two', args=[vpid1])`) deserializes the arguments, finds `vpid1` in `slotToVal`, and thus converts it into that earlier Promise. So if userspace in `two(argp)` does an eventual-send to the `argp` it received, that will be queued in the JS engine on `returnedP`. If `one` resolves it to a local object, those messages will be delivered locally, without kernel involvement.

If/when we implement redirects/forwards, and `one` forwards it to something imported from the kernel, then we'll need to arrange for a `syscall.send` to happen.